### DISCUSSION TITLE
Improving buildcache security: mirrors not URLs

### DISCUSSION THREAD

**Discussion by blue42u** [spack/spack#32958] [Ideas]
Buildcaches have a number of micro-issues (security and otherwise) that all stem from the same core design flaw, Spack allows no configuration for used buildcaches outside of their URL. Using [Spack in CI](https://github.com/spack/spack/discussions/32749) often aggravates these issues. This is not sustainable as Spack continues to roll forward and encourage using buildcaches in more scenarios.

## Observed issues
 - Buildcaches are fundamentally divorced from their signing keys. If any one buildcache is unsigned, **every** `spack install` that uses it needs to include `--no-check-signature`. This significantly weakens security for the buildcaches that are signed and encourages highly insecure behavior in users.
 
 - Buildcaches share a single set of signing keys, which are not pruned when buildcaches are removed from the `mirrors.yaml`. This means a "cracked" buildcache signing key could be used to maliciously inject arbitrary code via a user's current mirror set, *even if they aren't using the compromised buildcache anymore*. Although good keys are hard to crack, very few of us (users) are security professionals, trusting us to properly manage signing keys is not a good idea. As such, this is highly insecure behavior. (Debian [deprecated `apt-key` for basically this reason](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=851774#31).)
 
 - Buildcaches are fundamentally divorced from their access credentials. For buildcaches hosted in the cloud (e.g. AWS S3) in particular, this meant all the buildcaches needed to share a single set of credentials. If one buildcache was private while others were public, the credentials would need to have access to all buckets, breaking the principle of least privilege. Multiple private buildcaches would need to all be accessible via a single set of access credentials. A user would be charged if a buildcache was maliciously converted to a ["requester pays" bucket](https://docs.aws.amazon.com/AmazonS3/latest/userguide/RequesterPaysBuckets.html). This was highly insecure behavior and was fixed by https://github.com/spack/spack/pull/31391, but since it deeply relates I bring it up again here.
 
 - Buildcaches blindly and aggressively probe URLs. In a practical pipeline this means thousands of requests, almost all of which result in 404s (see image below). AFAICT this expensive behavior is unnecessary, some likely relates to [Spack's blind handling for deprecated and un/signed buildcache layouts](https://github.com/spack/spack/blob/fa7407093ec957729308f95a1403da3b222ec797/lib/spack/spack/binary_distribution.py#L1309-L1313). This is detrimental behavior that could be fixed with a modicum of configuration.
   ![image](https://user-images.githubusercontent.com/17242663/193618151-91998e9e-8e57-4218-b2e9-8435364c5e60.png)

## Proposed solution

Buildcaches should always be mirrors, and never a bare URL. Specifically:
 - Each mirror specification in `mirrors.yaml` contains all the configuration needed to use the associated buildcache.
   * Each mirror includes the signing key(s) for the associated buildcache. Binary packages are checked with the per-buildcache keys.
   * Each mirror also includes the buildcache layout version, e.g. `1` might mean the current layout with `index.json`, `index.json.hash` and `*.spec.json`/`*.spec.json.sig`/`*.spack` files named by arch/package/hash. `spack install` never probes for files outside of the configured layout, if files appear to be missing `spack mirror update` (see below) will be recommended.
   * By default mirrors do not allow direct fetches, i.e. `spack install` assumes the `index.json` is up-to-date. This can be enabled on a per-mirror basis.
 - `spack mirror` provides user-facing commands to transparently manage buildcache subscriptions:
   * `spack mirror add` automagically fills in the buildcache configuration by probing the buildcache URL and downloading keys.
   * `spack mirror update` repeats the above and updates the configuration. If the configuration is ever found to be wrong during a `spack install` (e.g. signature verification failure), this command is suggested as a solution.
 - The buildcache management commands `spack buildcache` and `spack ci` always use mirror names instead of directories or URLs:
   * `spack buildcache keys` is removed (now handled by `spack mirror (add|update)`.
   * `spack buildcache (add|create|update-index|...)` only accept a mirror name for their destination, URLs and directories are not allowed.
   * `spack ci generate` requires a destination mirror (i.e. `--buildcache-destination` is required and uses a mirror name instead of a URL). This mirror is *not* added to the concrete environment by default (to prevent security leakage).
   * `spack ci rebuild` takes a destination mirror as it's only argument, which should (almost always) be added via `spack mirror add` in the `before_script`.

I think this could also significantly simplify `binary_distribution.py`, since it can pass around first-class `Mirror` objects instead of bare URLs.

**Comment by iarspider** [comment]
> `spack buildcache (add|create|update-index|...)` only accept a mirror name for their destination, URLs and directories are not allowed.

Spack's concretization is still far from perfect (which deserves a separate discussion), so often I have to rebuild my environment from scratch (otherwise Spack insists that everything is installed and it has nothing to do). To reduce the build time, I keep a local buildcache - basically a directory on mounted file system. I don't want to use web/s3 for that, since it's easier to remove and update a local file while you are polishing a recipe. So, at least for me, the removal of directory-based buildcache will be bad.

Also, directory-based buildcaches will be used for so called "online" systems, which reside in an air-gapped "technical" network - creating a simple HTTP server, let alone a S3 one, is problematic there.

**Reply by blue42u** [reply]
@iarspider There are directory-based mirrors as well, so I don't think this proposed change would hurt your use case. The main difference is that when you would run:
```console
$ spack buildcache create -u -d /my/cool/buildcache
$ spack buildcache update-index -d /my/cool/buildcache
```
You would instead run:
```console
$ spack mirror add --buildcache-unsigned my-buildcache /my/cool/buildcache
$ spack buildcache create -m my-buildcache
$ spack buildcache update-index -m my-buildcache
```

Obviously the `spack mirror add` only needs to be done once and can be saved to your environment(s). So you don't even have to remember the directory/URL and the signing mode (`-u` vs. `-k`) in most cases, just the mirror's identifier name.

**Reply by iarspider** [reply]
Ah. Thanks for the explanation!

**Comment by scottwittenburg** [comment]
>     * Buildcaches blindly and aggressively probe URLs. In a practical pipeline this means thousands of requests, almost all of which result in 404s (see image below). AFAICT this expensive behavior is unnecessary, some likely relates to [Spack's blind handling for deprecated and un/signed buildcache layouts](https://github.com/spack/spack/blob/fa7407093ec957729308f95a1403da3b222ec797/lib/spack/spack/binary_distribution.py#L1309-L1313).  
> 

@tgamblin Last week we discussed how "<at>spackbot rebuild everything" should really prefer to consume dependency binaries built in the pipeline instead binaries from develop.  And the approach we discussed was to simply place a higher priority on mirror precedence rather than always preferring signed json files.  So kind of (though it isn't quite that simple) reversing the looping order [here](https://github.com/spack/spack/blob/fa7407093ec957729308f95a1403da3b222ec797/lib/spack/spack/binary_distribution.py#L1309-L1313), a couple lines of code @blue42u singled out above as "blind handling for deprecated and un/signed buildcache layout formats".

So for the "rebuild everything" use case, we are basically proposing that the unsigned binary built in the pipeline is preferable to the signed one in the `develop` mirror.  And now that I'm starting to think about the points raised in this discussion, I wonder if taking that approach will give @blue42u the impression I have completely disregarded the points he has raised, as that approach could be viewed as a step in the opposite direction from what this discussion proposes.

Since I don't want "rebuild everything" to have to wait until everything here is addressed, I wonder if we shouldn't just go with the other alternative we discussed last week: just remove the main mirror from the configuration when we know we're doing a "rebuild everything" pipeline.

**Comment by scottwittenburg** [comment]
> Each mirror also includes the buildcache layout version, e.g. 1 might mean the current layout with index.json, index.json.hash and *.spec.json/*.spec.json.sig

Does this mean that buildcaches will have to be homogenous?  So no mixing of the old `spec.yaml` specs with  the newer `spec.json` or `spec.json.sig` specs?

**Reply by blue42u** [reply]
Correct, I would even go one step further and say that mixing any of those three is not allowed. In other words, `spack install` will fetch only:
 - `*.spec.json.sig` if the layout is the "newer" version and a signing key is provided,
 - `*.spec.json` if the layout is the "newer" version and the signing key is disabled, or
 - `*.spec.yaml` if the layout is the "older" version.

Where all of those conditions are gathered from the settings in the `mirrors.yaml`, not from probing the buildcache itself. This way every file that is fetched (if the configuration is correct) should succeed and not generate a 404. If a 404 comes up, then the configuration is faulty and `spack install` tries a different buildcache (and recommends `spack mirror update`).

Is there a compelling argument for heterogeneous buildcaches? So far I've only thought up two scenarios where that might be useful:
 - To allow a signed buildcache to also be accessed unsigned. But checking the signatures is wicked fast already, so I don't see any practical benefit to doing this. And this weakens the security on the user's end, so IMHO I'd rather disallow this and encourage good security behavior.
 - To allow old Spack versions to share a buildcache with a newer Spack. But the hashes and versions seem to change rapidly, so I don't see much benefit here (except for just after a layout version bump). I also have no idea how one would reliably maintain a single buildcache pushed to by multiple Spack versions (it sounds to me like a recipe for disaster!)

**Comment by scottwittenburg** [comment]
> By default mirrors do not allow direct fetches, i.e. spack install assumes the index.json is up-to-date

While this assumption would make life a lot better (really, a huge amount, in my opinion), it's kind of hard for me to imagine in practice.  Because a lot of mirrors are populated by pipelines where concurrent jobs would need to edit the index at one time, I don't see how it could work.  From time to time, we've discussed making mirrors "smarter".  Instead of brainless s3 buckets or directories in a file-system, if mirrors were REST endpoints (or similar), we could ensure with each push of a binary package, that the index reflects the true state of the mirror, at least in theory.  As it is now though, how do we prevent the inevitable race conditions between parallel jobs updating the index simultaneously?  Not to mention that with brainless mirrors, updating the index can take forever, because we have to iterate over all the items, fetching and ingesting each one into a database, before we can finally write the index back to the mirror.

Don't get me wrong though, as I stated initially, I think it's a great goal to be able to assume the index is always up to date.

**Reply by blue42u** [reply]
I should have emphasized the "by default." My thought is that each mirror would have a flag that indicates whether direct fetches are allowed for that mirror, with the default being "not allowed."

In a `spack ci` pipeline, you'd set this flag in the "destination" mirror and no others. That should handle this case while still keeping the 404 rate at almost 0 (since in a `spack ci` pipeline the dependencies should have been uploaded already).

**Reply by blue42u** [reply]
> As it is now though, how do we prevent the inevitable race conditions between parallel jobs updating the index simultaneously?

I feel like this could be solved with filesystem locks and/or serializing all `update-index` jobs in `spack ci` pipelines using [the `resource_group:` job field](https://docs.gitlab.com/ee/ci/yaml/index.html#resource_group).

**Comment by scottwittenburg** [comment]
> Buildcaches should always be mirrors, and never a bare URL

Do I understand correctly the above statement is aimed at making sure spack never accesses the contents of a mirror without all the context you propose should go along with that mirror?

**Reply by blue42u** [reply]
Exactly correct.

**Comment by scottwittenburg** [comment]
fyi @zackgalbreath @kwryankrattiger for visibility and in case you are interested to participate in this discussion.
